File: dpgmm.c
#include "dpgmm.h"


DPGMM *dpgmm_init(int dims,int stickCap){
	int i;
	DPGMM *r=calloc(1,sizeof(DPGMM));
	r->dims=dims;
	r->stickCap=stickCap;
	r->data=malloc(sizeof(double)*LIMIT_DATA*dims);
	r->numData=0;
	r->prior=gaussian_prior_init(dims);
	r->n=malloc(sizeof(GaussianPrior*)*stickCap);
	for(i=0;i<stickCap;i++)
		r->n[i]=gaussian_prior_init(dims);
	r->beta=gsl_vector_alloc(2);
	gsl_vector_set_all(r->beta,1.0);
	r->alpha=gsl_vector_alloc(2);
	gsl_vector_set_all(r->alpha,1.0);
	r->v=gsl_matrix_alloc(stickCap,2);
	gsl_matrix_set_all(r->v,1.0);
	r->z=NULL;
	r->skip=0;
	r->epsilon=1e-4;
	r->nT=calloc(1,sizeof(StudentT*)*stickCap);
	r->vExpLog=gsl_vector_alloc(stickCap);
	r->vExpNegLog=gsl_vector_alloc(stickCap);
	gsl_vector_set_all(r->vExpNegLog,-1.0);
	gsl_vector_set_all(r->vExpLog,-1.0);
	return r;
}
void dpgmm_release(DPGMM *ctx){
	int i;
	free(ctx->data);
	gsl_vector_free(ctx->vExpNegLog);
	gsl_vector_free(ctx->vExpLog);
	for(i=0;i<ctx->stickCap;i++){
		if(ctx->nT[i]) student_t_free(ctx->nT[i]);
	}
	free(ctx->nT);
	gaussian_prior_free(ctx->prior);
	student_t_free(ctx->priorT);
	gsl_matrix_free(ctx->v);
	gsl_vector_free(ctx->alpha);
	gsl_vector_free(ctx->beta);
	for(i=0;i<ctx->stickCap;i++)
		gaussian_prior_free(ctx->n[i]);
	free(ctx->n);
	gsl_matrix_free(ctx->z);
	free(ctx);
}
void dpgmm_add(DPGMM *ctx,const double *sample){
	memcpy(&ctx->data[ctx->numData*ctx->dims],sample,sizeof(double)*ctx->dims);
	ctx->numData++;
	if((ctx->numData%LIMIT_DATA)==0){
		ctx->data=realloc(ctx->data,sizeof(double)*(ctx->numData+LIMIT_DATA));
	}
}
int dpgmm_setPrior(DPGMM *ctx,gsl_vector* mean,gsl_matrix* cover,double* weight,double scale){
	gaussian_prior_reset(ctx->prior);
	gaussian_prior_addPrior(ctx->prior,mean,cover,weight);
	ctx->priorT=gaussian_prior_intProb(ctx->prior);
	return 1;
}
int dpgmm_setDefaultsPrior(DPGMM *ctx){
	int i,j;
	gsl_vector *means=gsl_vector_alloc(ctx->dims);
	gsl_matrix *cover=gsl_matrix_alloc(ctx->dims,ctx->dims);
	for(i=0;i<ctx->dims;i++){
		gsl_vector_set(means,i,gsl_stats_mean(&ctx->data[i],ctx->dims,ctx->numData));
	}
	for(i=0;i<ctx->dims;i++){
		gsl_matrix_set(cover,i,i,gsl_stats_variance_m(&ctx->data[i],ctx->dims,ctx->numData,gsl_vector_get(means,i)));
		for(j=i+1;j<ctx->dims;j++){
			gsl_matrix_set(cover,i,j,gsl_stats_covariance_m(&ctx->data[i],ctx->dims,&ctx->data[j],ctx->dims,ctx->numData,gsl_vector_get(means,i),gsl_vector_get(means,j)));
			gsl_matrix_set(cover,j,i,gsl_matrix_get(cover,i,j));
		}
	}
	dpgmm_setPrior(ctx,means,cover,NULL,1);
	gsl_vector_free(means);
	gsl_matrix_free(cover);
	return 1;
}
double *dpgmm_getDM(DPGMM *ctx){
	return ctx->data;
}
double dpgmm_prob(DPGMM *ctx,const double *x){
	double ret,stick;
	int i;

	ret=0.0;
	stick=1.0;
	gsl_vector *vec=gsl_vector_alloc(ctx->v->size2);
	for(i=0;i<ctx->stickCap;i++){
		double bp=student_t_prob(ctx->nT[i],x);
		gsl_matrix_get_row(vec,ctx->v,i);
		double ev=gsl_matrix_get(ctx->v,i,0)/gsl_vector_sum(vec);
		ret += bp * stick * ev;
	        stick *= 1.0 - ev;
	}
	double bp=student_t_prob(ctx->priorT,x);
	ret += bp * stick;
	gsl_vector_free(vec);
	return ret;
}
int dpgmm_solv(DPGMM *ctx,int limitIter){
	gsl_matrix *newZ;
	double *dm=dpgmm_getDM(ctx);
	double *alpha,*theta,expLogStick,expNegLogStick;
	int offset,i,j;
	gsl_rng *r = gsl_rng_alloc (gsl_rng_default);
	if(ctx->z==NULL || ctx->z->size1<ctx->numData){
		newZ=gsl_matrix_alloc(ctx->numData,ctx->stickCap);
		if(ctx->z==NULL) offset=0;
		else{
			offset=ctx->z->size1;
			gsl_vector *v=gsl_vector_alloc(ctx->z->size2);
			for(i=0;i<offset;i++){	/*TODO:gsl matrix view*/
				gsl_matrix_get_row(v,ctx->z,i);
				gsl_matrix_set_row(newZ,i,v);
			}
			gsl_vector_free(v);
		}
		ctx->z=newZ;
		
		alpha=malloc(sizeof(double)*ctx->stickCap);
		theta=malloc(sizeof(double)*ctx->stickCap);
		for(i=0;i<ctx->stickCap;i++){
			alpha[i]=32.0;
		}

		if(ctx->stickCap==1){
			gsl_matrix_set_all(ctx->z,1.0);
		}else{
			size_t size=ctx->z->size1-offset;
			for(i=0;i<size;i++){
				gsl_ran_dirichlet(r,ctx->stickCap,alpha,theta);
				for(j=0;j<ctx->stickCap;j++){
					gsl_matrix_set(ctx->z,i,j,theta[j]);
				}
			}
		}
		free(alpha);
		free(theta);
	}
	gsl_rng_free(r);
	gsl_matrix *prev=gsl_matrix_clone(ctx->z);
	int iters=0;
	do{
		gsl_vector_set(ctx->alpha,0,gsl_vector_get(ctx->beta,0)+ctx->stickCap);
		gsl_vector_set(ctx->alpha,1,gsl_vector_get(ctx->beta,1)-gsl_vector_sum(ctx->vExpNegLog));
		double alphaRate=gsl_vector_get(ctx->alpha,0)/gsl_vector_get(ctx->alpha,1);
		expLogStick=-gsl_sf_psi(1.0 + alphaRate);
		expNegLogStick=expLogStick;
		expLogStick += gsl_sf_psi(1.0);
		expNegLogStick +=gsl_sf_psi(alphaRate);

		for(i=0;i<ctx->v->size1;i++){
			gsl_matrix_set(ctx->v,i,0,1.0);
			gsl_matrix_set(ctx->v,i,1,alphaRate);
		}

		gsl_vector *sums = gsl_matrix_sum_row(ctx->z);
		gsl_vector *cumsums = gsl_cumsum(sums);
		for(i=0;i<ctx->v->size1;i++){
			gsl_matrix_set(ctx->v,i,0,gsl_matrix_get(ctx->v,i,0)+gsl_vector_get(sums,i));
			gsl_matrix_set(ctx->v,i,1,gsl_matrix_get(ctx->v,i,1)+ctx->z->size1);
			gsl_matrix_set(ctx->v,i,1,gsl_matrix_get(ctx->v,i,1)-gsl_vector_get(cumsums,i));
		}
		gsl_vector_free(cumsums);
		gsl_vector_free(sums);
		for(i=0;i<ctx->v->size1;i++){
			double total=0.0;
			for(j=0;j<ctx->v->size2;j++){
				total+=gsl_matrix_get(ctx->v,i,j);
			}
			gsl_vector_set(ctx->vExpLog,i,-gsl_sf_psi(total));
		}
		
		gsl_vector_memcpy(ctx->vExpNegLog,ctx->vExpLog);
		
		for(i=0;i<ctx->v->size1;i++){
			gsl_vector_set(ctx->vExpLog,i,gsl_vector_get(ctx->vExpLog,i)+gsl_sf_psi(gsl_matrix_get(ctx->v,i,0)));
			gsl_vector_set(ctx->vExpNegLog,i,gsl_vector_get(ctx->vExpNegLog,i)+gsl_sf_psi(gsl_matrix_get(ctx->v,i,1)));
		}
		for(i=0;i<ctx->stickCap;i++){
			gaussian_prior_reset(ctx->n[i]);
			gaussian_prior_addGP(ctx->n[i],ctx->prior);
			double *weight=malloc(sizeof(double)*ctx->z->size1);
			
			for(j=0;j<ctx->z->size1;j++){
				weight[j]=gsl_matrix_get(ctx->z,j,i);
			}
			gaussian_prior_addSamples(ctx->n[i],dm,ctx->numData,weight);
			if(ctx->nT[i]) student_t_free(ctx->nT[i]);
			ctx->nT[i]=gaussian_prior_intProb(ctx->n[i]);
			free(weight);
		}
		
		gsl_vector *v=gsl_vector_alloc(ctx->z->size2);/*TODO:gsl matrix view*/
		for(i=ctx->skip;i<prev->size1;i++){
			gsl_matrix_get_row(v,ctx->z,i);
			gsl_matrix_set_row(prev,i,v);
		}
		gsl_vector_free(v);
		
		gsl_vector *vExpNegLogCum=gsl_cumsum(ctx->vExpNegLog);
		gsl_vector *base=gsl_vector_clone(ctx->vExpLog);
		
		for(i=1;i<vExpNegLogCum->size;i++){
			gsl_vector_set(base,i,gsl_vector_get(base,i)+gsl_vector_get(vExpNegLogCum,i-1));
		}
		
		gsl_vector *expTmp=gsl_vector_alloc(base->size);
		for(i=0;i<base->size;i++){
			gsl_vector_set(expTmp,i,exp(gsl_vector_get(base,i)));
		}
		
		for(i=ctx->skip;i<ctx->z->size1;i++){
			gsl_matrix_set_row(ctx->z,i,expTmp);
		}
		double *val=malloc(sizeof(double)*(ctx->numData-ctx->skip));
		for(i=0;i<ctx->stickCap;i++){
			bzero(val,sizeof(double)*(ctx->numData-ctx->skip));
			student_t_batchProb(ctx->nT[i],&dm[ctx->skip*ctx->dims],ctx->numData-ctx->skip,val);
			for(j=0;j<ctx->numData-ctx->skip;j++){
				gsl_matrix_set(ctx->z,ctx->skip+j,i,gsl_matrix_get(ctx->z,ctx->skip+j,i)*val[j]);
			}
		}
		
		student_t_batchProb(ctx->priorT,&dm[ctx->skip*ctx->dims],ctx->numData-ctx->skip,val);
		for(i=0;i<ctx->numData-ctx->skip;i++){
			val[i]*=exp(expLogStick + gsl_vector_get(vExpNegLogCum,vExpNegLogCum->size-1)) / (1.0 - exp(expNegLogStick));
		}
		double *diver=malloc(sizeof(double)*ctx->numData-ctx->skip*ctx->z->size2);
		for(i=0;i<ctx->numData-ctx->skip;i++){
			double total=0.0;
			for(j=0;j<ctx->z->size2;j++){
				total+=gsl_matrix_get(ctx->z,i+ctx->skip,j);
			}
			diver[i]=total+val[i];
		}
		for(i=0;i<ctx->z->size1-ctx->skip;i++){
			for(j=0;j<ctx->z->size2;j++){
				gsl_matrix_set(ctx->z,i+ctx->skip,j,gsl_matrix_get(ctx->z,i+ctx->skip,j)/diver[i]);
			}
		}
		
		double change=0.0;
		for(i=0;i<ctx->z->size1-ctx->skip;i++){
			double total=0.0;
			for(j=0;j<ctx->z->size2;j++){
				
				total+=fabs(gsl_matrix_get(prev,i+ctx->skip,j)-gsl_matrix_get(ctx->z,i+ctx->skip,j));
			}
			if(change<total) change=total;
		}
		gsl_vector_free(expTmp);
		gsl_vector_free(base);
		free(diver);
		free(val);
		gsl_vector_free(vExpNegLogCum);
		if(change<ctx->epsilon) break;
	}while(++iters<limitIter);

	gsl_matrix_free(prev);
	return iters;
	
}
/* TODO:double のNULLを定義*/

File: dpgmm.py
# Copyright 2011 Tom SF Haines

# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at

#   http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.



import math
import random
import numpy
import numpy.random
import scipy.special
import sys
from gcp import gcp



class DPGMM:
  def __init__(self, dims, stickCap = 1):
    self.dims = dims
    self.stickCap = stickCap

    self.data = [] # A list containing data matrices - used to collate all the samples ready for processing. Before processing they are all appended into a single data matrix, such that this list is of length 1.

    self.prior = gcp.GaussianPrior(self.dims) # The prior over the mixture components.
    self.priorT = None
    self.n = map(lambda _: gcp.GaussianPrior(self.dims), xrange(self.stickCap)) # The mixture components, one for each stick.
    self.beta = numpy.ones(2, dtype=numpy.float32) # The two parameters (Typically named alpha & beta) for the Gamma distribution prior over alpha.
    self.alpha = numpy.ones(2, dtype=numpy.float32) # The parameters for the Gamma distribution that represents the current distribution over alpha - basically beta updated with the current stick configuration.
    self.v = numpy.ones((self.stickCap,2), dtype=numpy.float32) # Each [i,:] of this array represents the two parameters of the beta distribution over the strick breaking weight for the relevant mixture component.
    self.z = None # The matrix of multinomials over stick-assignment for each sample, aligned with the data matrix. In the case of incrimental use will not necessarily be complete.

    self.skip = 0 # Number of samples at the start of the data matrix to not bother updating - useful to speed things up with incrimental learning.
    self.epsilon = 1e-4 # Amount of change below which it stops iterating.

    # The cache of stuff kept around for speed...
    self.nT = [None]*self.stickCap # The student T distribution associated with each Gaussian.
    self.vExpLog = numpy.empty(self.stickCap, dtype=numpy.float32) # The expected value of the logorithm of each v.
    self.vExpNegLog = numpy.empty(self.stickCap, dtype=numpy.float32) # The expected value of the logarithm of 1 minus each v.

    self.vExpLog[:] = -1.0    # Need these to always be correct - this matches initialisation of v.
    self.vExpNegLog[:] = -1.0 # As above.




  def setPrior(self, mean = None, covar = None, weight = None, scale = 1.0):
    """Sets a prior for the mixture components - basically a pass through for the addPrior method of the GaussianPrior class. If None (The default) is provided for the mean or the covar then it calculates these values for the currently contained sample set and uses them. Note that the prior defaults to nothing - this must be called before fitting the model, and if mean/covar are not provided then there must be enough data points to avoid problems. weight defaults to the number of dimensions if not specified. If covar is not given then scale is a multiplier for the covariance matrix - setting it high will soften the prior up and make it consider softer solutions when given less data. Returns True on success, False on failure - failure can happen if there is not enough data contained for automatic calculation (Think singular covariance matrix). This must be called before any solve methods are called."""
    # Handle mean/covar being None...
    if mean==None or covar==None:
      inc = gcp.GaussianInc(self.dims)
      dm = self.getDM()
      for i in xrange(dm.shape[0]): inc.add(dm[i,:])
      ggd = inc.fetch()
      if mean==None: mean = ggd.getMean()
      if covar==None: covar = ggd.getCovariance() * scale
      #print mean
      #print covar
      #mean(dm)

    if numpy.linalg.det(covar)<1e-12: return False

    # Update the prior...
    self.prior.reset()
    self.prior.addPrior(mean, covar, weight)
    self.priorT = self.prior.intProb()

    return True




  def add(self, sample):
    """Adds either a single sample or several samples - either give a single sample as a 1D array or a 2D array as a data matrix, where each sample is [i,:]. (Sample = feature. I refer to them as samples as that more effectivly matches the concept of this modeling the probability distribution from which the features are drawn.)"""
    sample = numpy.asarray(sample, dtype=numpy.float32)
    if len(sample.shape)==1:
      self.data.append(numpy.reshape(sample, (1,self.dims)))
    else:
      assert(len(sample.shape)==2)
      assert(sampler.shape[1]==self.dims)
      self.data.append(sample)

  def getDM(self):
    """Returns a data matrix containing all the samples that have been added."""
    if len(self.data)==1: return self.data[0]
    if len(self.data)==0: return None

    self.data = [numpy.vstack(self.data)]
    
    return self.data[0]

  def size(self):
    """Returns the number of samples that have been added."""
    dm = self.getDM()
    if dm!=None: return dm.shape[0]
    else: return 0


  def lock(self, num=0):
    """Prevents the algorithm updating the component weighting for the first num samples in the database - potentially useful for incrimental learning if in a rush. If set to 0, the default, everything is updated."""
    self.skip = num

  def solve(self, iterCap=None):
    """Iterates updating the parameters until the model has converged. Note that the system is designed such that you can converge, add more samples, and converge again, i.e. incrimental learning. Alternativly you can converge, add more sticks, and then convegre again without issue, which makes finding the correct number of sticks computationally reasonable.. Returns the number of iterations required to acheive convergance. You can optionally provide a cap on the number of iterations it will perform."""

   # Deal with the z array being incomplete - enlarge/create as needed. Random initialisation is used...
    dm = self.getDM()
    newZ = numpy.empty((dm.shape[0],self.stickCap), dtype=numpy.float32)
      
    offset = 0
    self.z = newZ

    self.z[offset:,:] = numpy.random.mtrand.dirichlet(32.0*numpy.ones(self.stickCap), size=self.z.shape[0]-offset) # 32 is to avoid extreme values, which can lock it in place, without the distribution being too flat as to cause problems converging.
    f = open('mat.txt', 'w')
    for i in range(self.z.shape[0]):
      for j in range(self.z.shape[1]):
        f.write(str(self.z[i,j])+' ')
      f.write('\n')
    f.close()
    # Iterate until convergance...
    prev = self.z.copy()
    iters = 0
    while True:
      iters += 1

      # Update the concentration parameter...
      self.alpha[0] = self.beta[0] + self.stickCap
      self.alpha[1] = self.beta[1] - self.vExpNegLog.sum()

      # Record the expected values of a stick given the prior alone - needed to normalise the z values...
      expLogStick = -scipy.special.psi(1.0 + self.alpha[0]/self.alpha[1])
      expNegLogStick = expLogStick
      expLogStick += scipy.special.psi(1.0)
      expNegLogStick += scipy.special.psi(self.alpha[0]/self.alpha[1])

      # Update the stick breaking weights...
      self.v[:,0] = 1.0
      self.v[:,1] = self.alpha[0]/self.alpha[1]
      
      sums = self.z.sum(axis=0)
      self.v[:,0] += sums
      
      self.v[:,1] += self.z.shape[0]
      self.v[:,1] -= numpy.cumsum(sums)
      # Calculate the log expectations on the stick breaking weights...
      self.vExpLog[:] = -scipy.special.psi(self.v.sum(axis=1))
      self.vExpNegLog[:] = self.vExpLog
      
      self.vExpLog[:] += scipy.special.psi(self.v[:,0])
      self.vExpNegLog[:] += scipy.special.psi(self.v[:,1])
     
      # Update the Gaussian conjugate priors, extracting the student-t distributions as well...
      #print sys.stdout.write(str(self.z))
      #for i in self.z:
      #  print i
      for k in xrange(self.stickCap):
        self.n[k].reset()
        self.n[k].addGP(self.prior)
        self.n[k].addSamples(dm, self.z[:,k])
#        print "self.n[k].mu:%f"%(self.n[k].mu)
        self.nT[k] = self.n[k].intProb()

      # Update the z values...
      prev[self.skip:,:] = self.z[self.skip:,:]
      
      vExpNegLogCum = self.vExpNegLog.cumsum()
      base = self.vExpLog.copy()
      
      base[1:] += vExpNegLogCum[:-1]
      
      self.z[self.skip:,:] = numpy.exp(base).reshape((1,self.stickCap))
      
      for k in xrange(self.stickCap):
        self.z[self.skip:,k] *= self.nT[k].batchProb(dm[self.skip:,:])

      norm = self.priorT.batchProb(dm[self.skip:,:])
      norm *= math.exp(expLogStick + vExpNegLogCum[-1]) / (1.0 - math.exp(expNegLogStick))

      self.z[self.skip:,:] /= (self.z[self.skip:,:].sum(axis=1) + norm).reshape((self.z.shape[0]-self.skip,1))

      # Check for convergance...
      change = numpy.abs(prev[self.skip:,:]-self.z[self.skip:,:]).sum(axis=1).max()
      print "iters:%d self.epsilon:%.10lf change:%.10lf"%(iters,self.epsilon,change)
      if change<self.epsilon: break
      if iters==iterCap: break

    # Return the number of iterations that were required to acheive convergance...
    return iters


  def sampleMixture(self):
    """Once solve has been called and a distribution over models determined this allows you to draw a specific model. Returns a 2-tuple, where the first entry is an array of weights and the second entry a list of Gaussian distributions - they line up, to give a specific Gaussian mixture model. For density estimation the probability of a specific point is then the sum of each weight multiplied by the probability of it comming from the associated Gaussian. For clustering the probability of a specific point belonging to a cluster is the weight multiplied by the probability of it comming from a specific Gaussian, normalised for all clusters. Note that this includes an additional term to cover the infinite number of terms that follow, which is really an approximation, but tends to be such a small amount as to not matter. Be warned that if doing clustering a point could be assigned to this 'null cluster', indicating that the model thinks the point belongs to an unknown cluster (i.e. one that it doesn't have enough information, or possibly sticks, to instanciate.)."""
    weight = numpy.empty(self.stickCap+1, dtype=numpy.float32)
    stick = 1.0
    for i in xrange(self.stickCap):
      val = random.betavariate(self.v[i,0], self.v[i,1])
      weight[i] = stick * val
      stick *= 1.0 - val
    weight[-1] = stick

    gauss = map(lambda x: x.sample(), self.n)
    gauss.append(self.prior.sample())
    
    return (weight,gauss)

  def intMixture(self):
    """Returns the details needed to calculate the probability of a point given the model (density estimation), or its probability of belonging to each stick (clustering), but with the actual draw of a mixture model from the model integrated out. It is an apprximation, though not a bad one. Basically you get a 2-tuple - the first entry is an array of weights, the second a list of student-t distributions. The weights and distributions align, such that for density estimation the probability for a point is the sum over all entrys of the weight multiplied by the probability of the sample comming from the student-t distribution. The prob method of this class calculates the use of this for a sample directly. For clustering the probability of belonging to each cluster is calculated as the weight multiplied by the probability of comming from the associated student-t, noting that you will need to normalise. stickProb allows you to get this assesment directly. Do not edit the returned value; also, it will not persist if solve is called again. This must only be called after solve is called at least once. Note that an extra element is included to cover the remainder of the infinite number of elements - be warned that a sample could have the highest probability of belonging to this dummy element, indicating that it probably belongs to something for which there is not enough data to infer a reasonable model."""
    weights = numpy.empty(self.stickCap+1, dtype=numpy.float32)
    
    stick = 1.0
    for i in xrange(self.stickCap):
      ev = self.v[i,0] / self.v[i,:].sum()
      weights[i] = stick * ev
      stick *= 1.0 - ev
    weights[-1] = stick
      
    return (weights, self.nT + [self.priorT])
  
  def prob(self, x):
    """Given a sample this returns its probability, with the actual draw from the model integrated out. Must not be called until after solve has been called. This is the density estimate if using this model for density estimation. Will also accept a data matrix, in which case it will return a 1D array of probabilities aligning with the input data matrix."""
    x = numpy.asarray(x)
    if len(x.shape)==1:
      ret = 0.0
      stick = 1.0
      for i in xrange(self.stickCap):
        bp = self.nT[i].prob(x)
        ev = self.v[i,0] / self.v[i,:].sum()
        ret += bp * stick * ev
        stick *= 1.0 - ev
      bp = self.priorT.prob(x)
      ret += bp * stick
      return ret
    else:
      ret = numpy.zeros(x.shape[0])
      stick = 1.0
      for i in xrange(self.stickCap):
        bp = self.nT[i].batchProb(x)
        ev = self.v[i,0] / self.v[i,:].sum()
        ret += bp * stick * ev
        stick *= 1.0 - ev
      bp = self.priorT.batchProb(x)
      ret += bp * stick
      return ret
  
  def stickProb(self, x):
    """Given a sample this returns its probability of belonging to each of the components, as a 1D array, including a dummy element at the end to cover the infinite number of sticks not being explicitly modeled. This is the probability of belonging to each cluster if using the model for clustering. Must not be called until after solve has been called. Will also accept a data matrix, in which case it will return a matrix with a row for each vector in the input data matrix."""
    x = numpy.asarray(x)
    if len(x.shape)==1:
      ret = numpy.empty(self.stickCap+1, dtype=numpy.float32)
      stick = 1.0
      for i in xrange(self.stickCap):
        bp = self.nT[i].prob(x)
        ev = self.v[i,0] / self.v[i,:].sum()
        ret[i] = bp * stick * ev
        stick *= 1.0 - ev
      bp = self.priorT.prob(x)
      ret[self.stickCap] = bp * stick
      ret /= ret.sum()
      return ret
    else:
      ret = numpy.empty((x.shape[0],self.stickCap+1), dtype=numpy.float32)
      stick = 1.0
      for i in xrange(self.stickCap):
        bp = self.nT[i].batchProb(x)
        ev = self.v[i,0] / self.v[i,:].sum()
        ret[:,i] = bp * stick * ev
        stick *= 1.0 - ev
      bp = self.priorT.batchProb(x)
      ret[:,self.stickCap] = bp * stick
      ret /= ret.sum(axis=1).reshape((-1,1))
      return ret


  def reset(self, alphaParam = True, vParam = True, zParam = True):
    """Allows you to reset the parameters associated with each variable - good for doing a restart if your doing multiple restarts, or if you suspect it has got stuck in a local minima whilst doing incrimental stuff."""
    if alphaParam:
      self.alpha[:] = 1.0

    if vParam:
      self.v[:,:] = 1.0
      self.vExpLog[:] = -1.0
      self.vExpNegLog[:] = -1.0

    if zParam:
      self.z = None


